sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -230,25 +229,43 @@ public void visitPrimitiveTransform(Node node) {
 
   private void replace(final PTransformOverride override) {
     final Collection<Node> matches = new ArrayList<>();
+    final Set<String> freedNames = new HashSet<>();
     transforms.visit(
         new PipelineVisitor.Defaults() {
+          Node clearingNamesFor = null;
+
           @Override
           public CompositeBehavior enterCompositeTransform(Node node) {
+            if (clearingNamesFor != null) {
+              freedNames.add(node.getFullName());
+              return CompositeBehavior.ENTER_TRANSFORM;
+            }
             if (!node.isRootNode() && override.getMatcher().matches(node.toAppliedPTransform())) {
               matches.add(node);
               // This node will be replaced. It should not be visited.
-              return CompositeBehavior.DO_NOT_ENTER_TRANSFORM;
+              clearingNamesFor = node;
+              freedNames.add(node.getFullName());
```

Comment:
It seems simpler and clearer to me to preserve the existing behavior (don't enter child).

Instead, add another function that is "collect all names to be replaces" that does an inner recursive traverse of the node to be replaced.

Traverses that do two things in modes based on local variables being null or not is ... confusing.